File: sharp_brain/sharp_brain/Neural/NeuralNet.cs
using SharpBrain.Activations;
using SharpBrain.Utils;
using System;
using System.Collections.Generic;
using System.Linq;

namespace SharpBrain.Neural
{
    public class NeuralNet
    {
        private readonly ActivationStrategy strategy;
        private readonly Activation activation;

        private int numInputs;
        private int numHiddenLayers;
        private List<int> numHidden;
        private int numOutput;
        private Layer inputLayer;
        private List<Layer> hiddenLayers;
        private Layer outputLayer;
        private Neuron bias;
        private int activationIndex = -1;

        public NeuralNet(List<int> topology, ActivationStrategy strategy)
        {
            this.strategy = strategy;
            SetupTopology(topology);
            Setup();
        }

        public NeuralNet(List<int> topology, Activation activation)
        {
            this.activation = activation;
            SetupTopology(topology);
            Setup();
        }

        public List<double> Predict(List<double> input)
        {
            var index = 0;
            foreach (var n in inputLayer)
            {
                n.Value = input[index++];
            }
            var output = new List<double>();
            foreach (var n in outputLayer)
            {
                output.Add(n.Activate());
            }
            return output;
        }

        public void Train(List<List<double>> input, List<List<double>> expected, int iterations = 500)
        {
            for (int i = 0; i < iterations; ++i)
            {
                var numSamples = input.Count;
                for (int j = 0; j < numSamples; ++j)
                {
                    Predict(input[j]);
                    for (int k = 0; k < numOutput; ++k)
                    {
                        var o = outputLayer[k] as OutputNeuron;
                        o.CalculateDelta(expected[j][k]);
                    }

                    for (int k = numHiddenLayers - 1; k >= 0; --k)
                    {
                        var currentLayer = hiddenLayers[k];
                        for (int hk = 0; hk < currentLayer.Count; ++hk)
                        {
                            var h = currentLayer[hk] as HiddenNeuron;
                            h.CalculateDelta(k == numHiddenLayers - 1 ? outputLayer : hiddenLayers[k + 1]);
                        }
                    }

                    for (int k = 0; k < numOutput; ++k)
                    {
                        var o = outputLayer[k] as OutputNeuron;
                        o.ApplyDelta();
                    }

                    for (int k = numHiddenLayers - 1; k >= 0; --k)
                    {
                        var currentLayer = hiddenLayers[k];
                        for (int hk = 0; hk < currentLayer.Count; ++hk)
                        {
                            var h = currentLayer[hk] as HiddenNeuron;
                            h.ApplyDelta();
                        }
                    }
                }
            }
        }

        private void SetupTopology(List<int> topology)
        {
            int lastIndex = topology.Count - 1;
            numInputs = topology[0];
            numOutput = topology[lastIndex];
            numHiddenLayers = lastIndex - 1;
            numHidden = topology.Slice(1, numHiddenLayers);
        }

        private void Setup()
        {
            inputLayer = new Layer();
            for (int i = 0; i < numInputs; ++i)
            {
                inputLayer.AddNeuron(new InputNeuron(0, i));
            }

            hiddenLayers = new List<Layer>();
            var layerIndex = 1;
            foreach (var numNeurons in numHidden)
            {
                var layer = new Layer();
                for (int i = 0; i < numNeurons; ++i)
                {
                    layer.AddNeuron(new HiddenNeuron(layerIndex, i, GetActivation(layerIndex)));
                }
                ++layerIndex;
                hiddenLayers.Add(layer);
            }

            outputLayer = new Layer();
            for (int i = 0; i < numOutput; ++i)
            {
                outputLayer.AddNeuron(new OutputNeuron(numHiddenLayers + 1, i, GetActivation(numHiddenLayers + 1)));
            }

            bias = new InputNeuron(-1, -1);
            bias.Value = 1.0;

            if (numHiddenLayers > 0)
            {
                foreach (var o in outputLayer)
                {
                    List<Connection> cons = new List<Connection>();
                    foreach (var n in hiddenLayers.Last())
                    {
                        cons.Add(new Connection(n));
                    }
                    cons.Add(new Connection(bias));
                    ((WorkingNeuron)o).Connections = cons;
                }

                foreach (var h in hiddenLayers[0])
                {
                    List<Connection> cons = new List<Connection>();
                    foreach (var n in inputLayer)
                    {
                        cons.Add(new Connection(n));
                    }
                    cons.Add(new Connection(bias));
                    ((WorkingNeuron)h).Connections = cons;
                }

                for (int i = 1; i < numHiddenLayers; ++i)
                {
                    foreach (var h in hiddenLayers[i])
                    {
                        List<Connection> cons = new List<Connection>();
                        foreach (var n in hiddenLayers[i - 1])
                        {
                            cons.Add(new Connection(n));
                        }
                        cons.Add(new Connection(bias));
                        ((WorkingNeuron)h).Connections = cons;
                    }
                }
            }
            else
            {
                foreach (var o in outputLayer)
                {
                    List<Connection> cons = new List<Connection>();
                    foreach (var n in inputLayer)
                    {
                        cons.Add(new Connection(n));
                    }
                    cons.Add(new Connection(bias));
                    ((WorkingNeuron)o).Connections = cons;
                }
            }
        }

        private Activation GetActivation(int index)
        {
            switch (strategy)
            {
                case ActivationStrategy.Random:
                    {
                        var r = new Random();
                        return Activation.Options[r.Next(0, Activation.Options.Count)];
                    }
                case ActivationStrategy.FixedRandom:
                    {
                        if (activationIndex == -1)
                        {
                            var r = new Random();
                            activationIndex = r.Next(0, Activation.Options.Count);
                        }
                        return Activation.Options[activationIndex];
                    }
                case ActivationStrategy.RoundRobin:
                    return Activation.Options[index % Activation.Options.Count];
                case ActivationStrategy.Fixed:
                    return activation;
                default:
                    return new Identity();
            }
        }
    }
}

File: java_brain/src/main/java/org/netbrains/javabrain/neural/NeuralNet.java
package org.netbrains.javabrain.neural;

import org.netbrains.javabrain.activation.Identity;
import org.netbrains.javabrain.activation.Sigmoid;

import java.util.ArrayList;
import java.util.List;

public class NeuralNet {
    private final int numInputs;
    private final int numHiddenLayers;
    private final int numOutput;
    private final List<Integer> numHidden;

    private List<Neuron> input;
    private List<List<WorkingNeuron>> hidden;
    private List<WorkingNeuron> output;

    public NeuralNet(List<Integer> topology) {
        numInputs = topology.get(0);
        numOutput = topology.get(topology.size() - 1);
        numHidden = new ArrayList<>();
        for(int i = 1; i < topology.size() - 1; ++i) {
            numHidden.add(topology.get(i));
        }
        numHiddenLayers = numHidden.size();
        setup();
    }

    public List<Double> predict(List<Double> inputs) {
        for(int i = 0; i < inputs.size(); ++i) {
            ((InputNeuron)input.get(i)).setValue(inputs.get(i));
        }
        List<Double> outputs = new ArrayList<>();
        for(WorkingNeuron o : output) {
            outputs.add(o.activate());
        }
        return outputs;
    }

    public void train(List<List<Double>> inputs, List<List<Double>> expected, int iterations) {
        for(int i = 0; i < iterations; ++i) {
            for(int j = 0; j < inputs.size(); ++j) {
                List<Double> in = inputs.get(j);
                List<Double> out = expected.get(j);

                List<Double> actual = predict(in);
                for(int k = 0; k < actual.size(); ++k) {
                    ((OutputNeuron)output.get(k)).calculateDelta(out.get(k));
                }

                for(int k = hidden.size() - 1; k >= 0; --k) {
                    for(int hk = 0; hk < hidden.get(k).size(); ++hk) {
                        List<WorkingNeuron> nextLayer = output;
                        ((HiddenNeuron)hidden.get(k).get(hk)).calculateDelta(nextLayer);
                    }
                }

                for(WorkingNeuron o : output) {
                    o.applyDelta();
                }

                for(List<WorkingNeuron> hl : hidden) {
                    for(WorkingNeuron h : hl) {
                        h.applyDelta();
                    }
                }
            }
        }
    }

    private void setup() {
        input = new ArrayList<>();
        for (int i = 0; i < numInputs; ++i) {
            input.add(new InputNeuron(0, i));
        }

        output = new ArrayList<>();
        for(int i = 0; i < numOutput; ++i) {
            output.add(new OutputNeuron(numHiddenLayers + 2, i, new Sigmoid()));
        }

        hidden = new ArrayList<>();
        int layerIndex = 1;
        for(int n : numHidden) {
            List<WorkingNeuron> hidden = new ArrayList<>();
            for(int i = 0; i < n; ++i) {
                hidden.add(new HiddenNeuron(layerIndex++, i, new Sigmoid()));
            }
            this.hidden.add(hidden);
        }

        if(numHiddenLayers > 0) {
            for (WorkingNeuron o : output) {
                List<Connection> connections = new ArrayList<>();
                for (WorkingNeuron h : hidden.get(hidden.size() - 1)) {
                    connections.add(new Connection(h));
                }
                connections.add(new Connection(new BiasNeuron(-1, -1,1.0)));
                o.setConnections(connections);
            }

            for (WorkingNeuron h : hidden.get(0)) {
                List<Connection> connections = new ArrayList<>();
                for (Neuron i : input) {
                    connections.add(new Connection(i));
                }
                connections.add(new Connection(new BiasNeuron(-1, -1,1.0)));
                h.setConnections(connections);
            }

            for (int i = 1; i < numHiddenLayers; ++i) {
                for (WorkingNeuron h1 : hidden.get(i)) {
                    List<Connection> connections = new ArrayList<>();
                    for (WorkingNeuron h2 : hidden.get(i - 1)) {
                        connections.add(new Connection(h2));
                    }
                    connections.add(new Connection(new BiasNeuron(-1, -1,1.0)));
                    h1.setConnections(connections);
                }
            }
        } else {
            for (WorkingNeuron o : output) {
                List<Connection> connections = new ArrayList<>();
                for (Neuron n : input) {
                    connections.add(new Connection(n));
                }
                connections.add(new Connection(new BiasNeuron(-1, -1,1.0)));
                o.setConnections(connections);
            }
        }
    }
}
